### PR TITLE
fix: `regression` end to end test

### LINKED ISSUE #1800: Could not satisfy all constraints for `regression` test
### Aim

Tried to prove and verify the `regression` test

### Expected Behavior

The test should produce a valid proof

### Bug

The test fails to prove and verify, outputtting the message: `Could not satisfy all constraints`

### To Reproduce

1.
2.
3.
4.

### Installation Method

None

### Nargo Version

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by kevaundray**
# Description

## Problem\*

Resolves #1800 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by kevaundray** [comment]
This is based off of #1951

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs [hunk 1]
```diff
@@ -767,10 +767,20 @@ impl GeneratedAcir {
...
+        // will be switched off.
         let mut expr = Expression::default();
         expr.push_addition_term(two_max_bits, q_witness);
         expr.push_addition_term(FieldElement::one(), r_witness);
-        self.push_opcode(AcirOpcode::Arithmetic(&comparison_evaluation - &expr));
+
+        let equation = &comparison_evaluation - &expr;
+        let equation_reduced: Expression = self.create_witness_for_expression(&equation).into();
+        let predicate_reduced: Expression = self.create_witness_for_expression(&predicate).into();
+        let predicated_equation = (&equation_reduced * &predicate_reduced).expect("operands were reduced to witnesses and so we expect the result of their multiplication to fit into a degree-2 expression");
```
Comment (TomAFrench) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs:781]:
```suggestion
        let equation = &comparison_evaluation - &expr;
        let predicated_equation = self.mul_with_witness(&equation, &predicate);
        self.push_opcode(AcirOpcode::Arithmetic(predicated_equation));
```

We have a helper function in `sort.rs` to do this kind of multiplication. This seems to add an extra 5 gates for some reason though.

Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs:781]: That sounds strange, let me take a look

File: crates/nargo_cli/tests/test_data_ssa_refactor/regression/src/main.nr [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+#[test]
+// Issue 1144
+fn test_1144()
+{
+    main([0x3f, 0x1c, 0xb8, 0x99, 0xab], 3);
+}
+
+// Issue 1169
+fn enc_test()
```
Comment (jfecher) [inline crates/nargo_cli/tests/test_data_ssa_refactor/regression/src/main.nr:92]: This function is never called

File: crates/nargo_cli/tests/test_data_ssa_refactor/regression/src/main.nr [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	let (nib, len) = compact_decode(x,z);
+	assert(len == 5);
+	assert([nib[0], nib[1], nib[2], nib[3], nib[4]] == [15, 1, 12, 11, 8]);
+
+
+}
+
+#[test]
+// Issue 1144
+fn test_1144()
```
Comment (jfecher) [inline crates/nargo_cli/tests/test_data_ssa_refactor/regression/src/main.nr:86]: What is our stance on `#[test]` in these test cases? We don't run `nargo test` on them so these tests aren't actually tested automatically by the CI.
For this case specifically, it seems we're already testing for it on line 76 so I think we can remove this function.

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs [hunk 4]
```diff
@@ -428,11 +428,22 @@ impl GeneratedAcir {
...
+        // Reduce the rhs_constraint to a witness
+        let rhs_constrain_reduced: Expression =
+            self.create_witness_for_expression(&rhs_constraint).into();
+        // Reduce the lhs_constraint to a witness
+        let lhs_reduced: Expression = self.create_witness_for_expression(lhs).into();
+
+        let div_euclidean = &(&lhs_reduced * predicate).expect(
+            "lhs_reduced should be a degree-1 witness and predicate should be a degree-1 witness",
+        ) - &(&rhs_constrain_reduced * predicate).expect(
+            "rhs_reduced should be a degree-1 witness and predicate should be a degree-1 witness",
```
Comment (TomAFrench) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs:445]: > predicate should be a degree-1 witness

Where is this enforced?

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs [hunk 5]
```diff
@@ -428,11 +428,22 @@ impl GeneratedAcir {
         // When the predicate is 0, the equation always passes.
         // When the predicate is 1, the euclidean division needs to be
         // true.
-        let mut rhs_constraint = (rhs * &Expression::from(q_witness)).unwrap();
+        let rhs_reduced: Expression = self.create_witness_for_expression(rhs).into();
+        let mut rhs_constraint = (&rhs_reduced * &Expression::from(q_witness))
+            .expect("rhs_reduced is expected to be a degree-1 witness");
         rhs_constraint = &rhs_constraint + r_witness;
```
Comment (TomAFrench) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs:434]:
```suggestion
        let rhs_constraint = &mul_with_witness(&rhs, &q_witness.into()) + r_witness;
```

**Comment by TomAFrench** [comment]
Addressed my review comments in https://github.com/noir-lang/noir/pull/1989 as the changes have already hit master.

File: crates/nargo_cli/tests/test_data_ssa_refactor/regression/src/main.nr (hunk 3 above)
Comment (kevaundray) [reply crates/nargo_cli/tests/test_data_ssa_refactor/regression/src/main.nr:86]: We probably don't want tests here as the default behavior is to not run tests, so they may get skipped

**Comment by kevaundray** [comment]
@guipublic can you also give an explicit approval for this PR!

**Review by TomAFrench — APPROVED** [review]
Lgtm

**Comment by TomAFrench** [comment]
Removing from queue as kev asked for guillaume to take a look. Feel free to kick off merge again.

**Review by guipublic — APPROVED** [review]

**Pull request merged**